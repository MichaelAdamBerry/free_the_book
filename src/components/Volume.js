import React from "react";
import queryString from "query-string";
import { getVolume } from "../api/google";
import CoverImg from "./CoverImg";

//TODO handle error if image doesn't exist

function VolumeView({ volume }) {
  const { volumeInfo, averageRating } = volume;
  const { title, authors, publishedDate, description, pageCount } = volumeInfo;

  console.log("authors ", authors);

  console.log("average rating is ", averageRating);
  return (
    <div>
      <h5>{title}</h5>
      <p>Published: {publishedDate}</p>
      <p>{pageCount}</p>
      <p>{averageRating}</p>
      <p dangerouslySetInnerHTML={{ __html: `${description}` }} />
      <CoverImg volumeInfo={volumeInfo} maxRes={true} />
    </div>
  );
}

export default class Volume extends React.Component {
  constructor(props) {
    super(props);
    this.state = { loading: true, volume: null };
  }
  async componentDidMount() {
    const id = queryString.parse(this.props.location.search);
    console.log("the id from the string is ", id.id);
    const volume = await getVolume(id.id);
    console.log("volume returned from api request", volume);
    this.setState({ volume: volume }, () => {
      this.setState({ loading: false });
    });
  }
  render() {
    const { loading, volume } = this.state;
    return <div>{!loading && <VolumeView volume={volume} />}</div>;
  }
}
