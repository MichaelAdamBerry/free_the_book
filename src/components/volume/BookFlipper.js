import React from "react";
import { Spring, config, animated as a } from "react-spring";
import CoverImg from "../CoverImg";
import PropTypes from "prop-types";
import stripHtml from "string-strip-html";

const Front = ({ imageLinks }) => {
  return <CoverImg imageLinks={imageLinks} maxRes={true} width={"350px"} />;
};

const Back = () => {
  return <div className="back" />;
};

export default class BookFlipper extends React.Component {
  state = { flipped: false };

  static propTypes = {
    description: PropTypes.string,
    previewLink: PropTypes.string,
    imageLinks: PropTypes.string
  };

  handleClick = () => {
    this.setState(prevState => ({ flipped: !prevState.flipped }));
  };

  render() {
    const { description, previewLink, imageLinks } = this.props;
    const { flipped } = this.state;
    return (
      <div onClick={this.handleClick} className="imgWrapper">
        <Spring
          config={key =>
            key === "opacity"
              ? { tension: "120", friction: "120" }
              : config.slow
          }
          to={{
            opacity: `${flipped ? "1" : "0"}`,
            transform: `perspective(600px) rotateY(${
              flipped ? "180" : "0"
            }deg)`,
            backgroundColor: `${flipped ? "#2f2f2f" : "#f7fbfc"}`
          }}>
          {({ opacity, transform }) => (
            <>
              <a.div
                className="c front"
                style={{
                  opacity: `${flipped ? "0" : "1"}`,
                  transform,
                  boxShadow: "rgb(51, 46, 46) 5px 5px 30px"
                }}>
                <Front imageLinks={imageLinks} />
              </a.div>
              <a.div
                className="c back"
                style={{
                  opacity,
                  transform,
                  border: "1.5px solid rgb(33, 33, 33)",
                  borderRadius: "3%",
                  boxShadow: "rgb(51, 46, 46) 5px 5px 30px"
                }}>
                <Back description={description} previewLink={previewLink} />
              </a.div>
              <a.div
                className="c"
                style={{
                  opacity
                }}>
                <p>{stripHtml(description)}</p>
              </a.div>
            </>
          )}
        </Spring>
      </div>
    );
  }
}
