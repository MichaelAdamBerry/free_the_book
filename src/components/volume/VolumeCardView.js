import React from "react";
import moment from "moment";
import PropTypes from "prop-types";
import Stars from "../card/Stars";
import BookFlipper from "./BookFlipper";

const Catagories = ({ categories }) => {
  return (
    <p>
      {categories.map(e => (
        <span key={Math.random() * 20}>{e} </span>
      ))}
    </p>
  );
};

Catagories.propTypes = {
  catagories: PropTypes.array
};

function VolumeCardView(props) {
  const {
    title,
    author,
    publishedDate,
    imageLinks,
    averageRating,
    description,
    pageCount,
    previewLink,
    categories
  } = props;
  return (
    <div className="volumeContainer">
      <div>
        <div className="miniHeader">
          <div className="titleInfo">
            <h5>{title}</h5>
            <p>{moment(publishedDate).format("Y")}</p>
            <div>
              {!averageRating === false && (
                <Stars averageRating={props.averageRating} />
              )}
            </div>
          </div>
          <div className="volumeAuthor">
            <h4>{author}</h4>
          </div>
          <div className="titleMeta">
            <p>{pageCount} pages</p>
            {!categories === false && <Catagories categories={categories} />}
          </div>
          <div>
            <BookFlipper
              imageLinks={imageLinks}
              description={description}
              previewLink={previewLink}
            />
          </div>
        </div>
      </div>
    </div>
  );
}

export default VolumeCardView;
