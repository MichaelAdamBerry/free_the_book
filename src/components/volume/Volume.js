import React from "react";
import queryString from "query-string";
import { getVolume } from "../../utils/api/google";
import { Spring } from "react-spring";
import Card from "../card/Card";
import VolumeCardView from "./VolumeCardView";
import PropTypes from "prop-types";
import Spinner from "../Spinner";

const RenderVolume = ({ volume }) => {
  return (
    <Spring from={{ opacity: 0 }} to={{ opacity: 1 }} delay={300}>
      {({ opacity }) => (
        <div style={{ opacity }}>
          <Card
            volume={volume}
            volumeInfo={volume.volumeInfo}
            render={({ props }) => {
              return <VolumeCardView {...props} />;
            }}
          />
        </div>
      )}
    </Spring>
  );
};

RenderVolume.propTypes = {
  volume: PropTypes.object.isRequired
};

export default class Volume extends React.Component {
  state = { loading: true };
  async componentDidMount() {
    const urlStr = queryString.parse(this.props.location.search);
    const volume = await getVolume(urlStr.id);
    this.setState({ volume });
    setTimeout(() => {
      this.setState({ loading: false });
    }, 1000);
  }
  render() {
    const { loading, volume } = this.state;
    return (
      <div>
        {!loading ? (
          <RenderVolume volume={volume} />
        ) : (
          <div className="volumeContainer" style={{ display: "flex" }}>
            <Spinner style={{ alignItems: "center" }} />
          </div>
        )}
      </div>
    );
  }
}
