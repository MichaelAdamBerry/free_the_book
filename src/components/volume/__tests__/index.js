import React from "react";
import renderer from "react-test-renderer";
import BookFlipper from "../BookFlipper";
import Volume from "../Volume";
import VolumeCardView from "../VolumeCardView";

const description =
  "There is a road no simple highway and if you go then you go alone";
const imageLinks = "https://i.ytimg.com/vi/671AgW9xSiA/hqdefault.jpg//";
const previewLinks = "https://google.com";

test("BookFlipper renders correct view when clicked", () => {
  const component = renderer.create(
    <BookFlipper
      description={description}
      imageLinks={imageLinks}
      previewLink={previewLinks}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  //flip card manually
  component.getInstance().handleClick();
  //rerender
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Volume Component Matches Snapshot", () => {
  const component = renderer.create(<Volume />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("VolumeCardView matches snapshot", () => {
  const component = renderer.create(<VolumeCardView />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
