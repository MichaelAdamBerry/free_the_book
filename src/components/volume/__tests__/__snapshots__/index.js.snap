// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookFlipper renders correct view when clicked 1`] = `
<div
  className="imgWrapper"
  onClick={[Function]}
>
  <div
    className="c front"
    style={
      Object {
        "boxShadow": "rgb(51, 46, 46) 5px 5px 30px",
        "opacity": "1",
        "transform": "perspective(600px) rotateY(0deg)",
      }
    }
  >
    <div
      style={
        Object {
          "height": "inherit",
          "overflow": "hidden",
        }
      }
    >
      <img
        alt="book cover"
        src="/"
        style={
          Object {
            "height": "auto",
            "width": "350px",
          }
        }
      />
    </div>
  </div>
  <div
    className="c back"
    style={
      Object {
        "border": "1.5px solid rgb(33, 33, 33)",
        "borderRadius": "3%",
        "boxShadow": "rgb(51, 46, 46) 5px 5px 30px",
        "opacity": "0",
        "transform": "perspective(600px) rotateY(0deg)",
      }
    }
  >
    <div
      className="back"
    />
  </div>
  <div
    className="c"
    style={
      Object {
        "opacity": "0",
      }
    }
  >
    <p>
      There is a road no simple highway and if you go then you go alone
    </p>
  </div>
</div>
`;

exports[`BookFlipper renders correct view when clicked 2`] = `
<div
  className="imgWrapper"
  onClick={[Function]}
>
  <div
    className="c front"
    style={
      Object {
        "boxShadow": "rgb(51, 46, 46) 5px 5px 30px",
        "opacity": "0",
        "transform": "perspective(600px) rotateY(0deg)",
      }
    }
  >
    <div
      style={
        Object {
          "height": "inherit",
          "overflow": "hidden",
        }
      }
    >
      <img
        alt="book cover"
        src="/"
        style={
          Object {
            "height": "auto",
            "width": "350px",
          }
        }
      />
    </div>
  </div>
  <div
    className="c back"
    style={
      Object {
        "border": "1.5px solid rgb(33, 33, 33)",
        "borderRadius": "3%",
        "boxShadow": "rgb(51, 46, 46) 5px 5px 30px",
        "opacity": "0",
        "transform": "perspective(600px) rotateY(0deg)",
      }
    }
  >
    <div
      className="back"
    />
  </div>
  <div
    className="c"
    style={
      Object {
        "opacity": "0",
      }
    }
  >
    <p>
      There is a road no simple highway and if you go then you go alone
    </p>
  </div>
</div>
`;

exports[`Volume Component Matches Snapshot 1`] = `<div />`;

exports[`VolumeCardView matches snapshot 1`] = `
<div
  className="volumeContainer"
>
  <div>
    <div
      className="miniHeader"
    >
      <div
        className="titleInfo"
      >
        <h5 />
        <p>
          2019
        </p>
        <div />
      </div>
      <div
        className="volumeAuthor"
      >
        <h4 />
      </div>
      <div
        className="titleMeta"
      >
        <p>
           pages
        </p>
      </div>
      <div>
        <div
          className="imgWrapper"
          onClick={[Function]}
        >
          <div
            className="c front"
            style={
              Object {
                "boxShadow": "rgb(51, 46, 46) 5px 5px 30px",
                "opacity": "1",
                "transform": "perspective(600px) rotateY(0deg)",
              }
            }
          >
            <div
              style={
                Object {
                  "height": "inherit",
                  "overflow": "hidden",
                }
              }
            >
              <img
                alt="book cover"
                src="http://www.lse.ac.uk/International-History/Images/Books/NoBookCover.png"
                style={
                  Object {
                    "height": "auto",
                    "width": "350px",
                  }
                }
              />
            </div>
          </div>
          <div
            className="c back"
            style={
              Object {
                "border": "1.5px solid rgb(33, 33, 33)",
                "borderRadius": "3%",
                "boxShadow": "rgb(51, 46, 46) 5px 5px 30px",
                "opacity": "0",
                "transform": "perspective(600px) rotateY(0deg)",
              }
            }
          >
            <div
              className="back"
            />
          </div>
          <div
            className="c"
            style={
              Object {
                "opacity": "0",
              }
            }
          >
            <p />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
